### PR TITLE
Fix handling of non-ASCII characters in archive entry file names

### LINKED ISSUE #12986: Repo rules fail to extract unicode archives due to latin-1 hack
### Description of the problem / feature request:

`rules_go` currently [fails on some machines](https://github.com/bazelbuild/rules_go/issues/2771) due to some [unicode characters](https://github.com/golang/go/tree/master/test/fixedbugs/issue27836.dir) included in filenames within the Go source archive - specifically the character [Ä](https://www.compart.com/en/unicode/U+00C4). For Linux and macOS, Go archives are distributed as [`tar.gz`](https://golang.org/dl/) files with [pax headers](https://en.wikipedia.org/wiki/Tar_(computing)#POSIX.1-2001/pax) in the tar files.

Affected systems include:
* ZFS volumes with the `utf8only` option. This is Ubuntu's default when choosing ZFS at install time.
* macOS (HFS+ and APFS require UTF-8 filenames)

Bazel uses Apache Commons Compress to extract tar archives. For most tar files, Commons Compress defers to the encoding specified by the JVM's `-Dfile.encoding` param, or the platform default. With ISO-8859-1 - Bazel's preference - UTF-8 encoded filename bytes in tar files basically pass through verbatim when extracted and everything works.

But when the tar entry has a pax `path` header, the path name is [always decoded as UTF-8](https://github.com/apache/commons-compress/blob/rel/1.19/src/main/java/org/apache/commons/compress/archivers/tar/TarArchiveInputStream.java#L516-L517). 

The character Ä in its [composed](https://en.wikipedia.org/wiki/Unicode_equivalence) form is unicode character `U+00C4`. In Java's internal UTF-16, it's simply represented as `0xC4`. But encoded as UTF-8 it becomes the multi-byte sequence `0xC3 0x84` since `0xC4` as a single byte is not a valid UTF-8 value.

When Commons Compress parses a pax-formatted tar file with a filename containing Ä as the `0xC3 0x84` UTF-8 string, the resulting Java string contains the value `0xC4` after decoding. But this value is never re-encoded as UTF-8 when creating the file on the filesystem. Instead, Bazel uses the Java char values [verbatim as long as they're < 0xff](https://github.com/bazelbuild/bazel/blob/master/src/main/native/latin1_jni_path.cc#L84). An attempt to create a filename containing `0xC4` on a filesystem that requires UTF-8 filenames will fail.

But `rules_go` doesn't currently fail on macOS systems despite them also requiring UTF-8 filenames. This is because the `darwin` archives use decomposed representations of unicode characters. OS X has a history of preferring the decomposed forms over composed.

So instead of Ä being `U+00C4` ("Latin Capital Letter A with Diaeresis"), it's `U+0041` (just capital A) followed by `U+0308` ("Combining Diaeresis"). Encoded in UTF-8 as seen in the macOS golang tarballs, the byte string is `0x41 0xCC 0x88`. Decoded to a Java string (16-bit chars) it's `0x0041 0x0308`. Coincidentally I presume, Bazel is able to extract this decomposed form on UTF-8 filesystems because it [ignores](https://github.com/bazelbuild/bazel/blob/master/src/main/native/latin1_jni_path.cc#L84) the diaeresis and replaces it with a literal `'?'` character. So instead of `Äfoo.go` as is contained in the Go source archive, Bazel writes `A?foo.go` on macOS.

Like Linux on ZFS, Bazel fails to extract the linux archive on macOS as reported [here](https://github.com/bazelbuild/rules_go/issues/2773).

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

`test.bzl`
```python
def _tar_round_trip_impl(ctx):
    ctx.file("Äfoo.txt", "boo!\n")
    ctx.execute(["tar", "--format=" + ctx.attr.format, "-czvf", "file.tar.gz", "Äfoo.txt"])
    ctx.extract("file.tar.gz", "out")
    ctx.file("BUILD.bazel", 'exports_files(["Äfoo.txt", "out/Äfoo.txt", "file.tar.gz"])', legacy_utf8=False)

tar_round_trip = repository_rule(
    implementation = _tar_round_trip_impl,
    attrs = {
	"format": attr.string(
            mandatory = True,
        ),
    },
)
```

`WORKSPACE`
```python
load("//:test.bzl", "tar_round_trip")

tar_round_trip(
    name = "non_pax",
    format = "ustar",  # supported by both macos BSD tar and linux GNU tar
)

tar_round_trip(
    name = "pax",
    format = "pax",
)
```

`BUILD`
```python
genrule(
    name = "non_pax_test",
    srcs = ["@non_pax//:out/Äfoo.txt"],
    outs = ["non_pax.txt"],
    cmd = """
        cp $(location @non_pax//:out/Äfoo.txt) "$@"
    """,
)

genrule(
    name = "pax_test",
    srcs = ["@pax//:out/Äfoo.txt"],
    outs = ["pax.txt"],
    cmd = """
	cp $(location @pax//:out/Äfoo.txt) "$@"
    """,
)
```

```shell
# Works
bazel build //:non_pax_test

# Fails, either due to not being able to write the file (utf8 filesystem),
# or because the written filename is mangled.
bazel build //:pax_test
```

### What operating system are you running Bazel on?
Mac OS 10.15.7
Ubuntu 20.04 with a ZFS root partition with `utf8only` enabled (the default for Ubuntu's ZFS support).

### What's the output of `bazel info release`?
`release 4.0.0`

###  Have you found anything relevant by searching the web?
https://github.com/bazelbuild/rules_go/issues/2771
https://github.com/bazelbuild/bazel/issues/374 - pretty generic issue regarding filename characters.
https://github.com/bazelbuild/bazel/issues/7055 - an issue with the same problematic file in the Go archive, but targeted at Darwin only.

### PR DESCRIPTION
**Pull request by fmeum**
When creating a `PathFragment` from a ZIP or TAR entry file name, the raw bytes of the name are now wrapped into a Latin-1 encoded String, which is how Bazel internally represents file paths.

Previously, ZIP entries as well as TAR entries with PAX headers would result in ordinary decoded Java strings, resulting in corrupted file names when passed to Bazel's file system operations.

Fixes #12986
Fixes https://github.com/bazelbuild/rules_go/issues/2771

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java
Comment (fmeum) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java:61]: There doesn't seem to be an easy way to keep reading legacy tars in Latin-1 while still being able to distinguish them from UTF-8 names. 

In order to preserve backwards compatibility with Latin-1 tars without PAX headers, I *could* inject a custom `Charset` via `CharsetProvider` that somehow marks all strings it produces in a way that lets us distinguish them from those created by the default UTF-8 charset (e.g. via a `WeakIdentityHashMap`). Since that would definitely be a hack, I didn't want to do it without getting feedback on this point first.

Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java:61]: cc @tetromino (since he's been thnking about Unicode recently)

My professional opinion is that I don't know. What would his hack be useful for? I thought that what this change was doing was to parse the UTF-8 data into proper Unicode then convert that to Bazel's weird "UTF-8 bytes in ISO8859-1" encoding, but that seems doable without resorting to somehow tagging the `String`s produced by it by (essentially) doing "if valid UTF-8, decode it as such, otherwise go directly to the weird Bazel encoding".

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java:61]: It's more messy than that: The Latin-1 trick only works if all strings are encoded in Latin-1 from source to sink, otherwise the "opaque bag of bytes" property is irrecoverably lost. 

But commons-compress breaks this invariant: While tar ustar headers respect the default string encoding (which can also be set via the constructor), the more modern tar pax headers always use UTF-8. 

Since Bazel can't determine which encoding was used to produce the file names returned by commons-compress, I found the only non-hacky option to be to force UTF-8 for ustar headers as well, decode UTF-8 and then reencode as Latin-1.

As a result, this change would break users of tars with non-UTF-8 ustar headers. Previously, these were handled correctly via the Latin-1 trick (but any pax header with non-ASCII chars were broken), but now they would be decoded as UTF-8 and thus turn into garbage.

Ideally we could tweak commons-compress to return byte arrays or at least tell us which type of header it parser, but I didn't find a way to do this. We could raise an issue with them, hack around this limitation by tracking the strings via a custom charset or accept that life isn't fair and UTF-8 deserves to be the winner. ;⁠-⁠)

Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java:61]: Well, actually, it's even more messy than that, if I understand correctly: the nice thing about the Latin-1 hack is that it works as long as the encoding on the input is the same as the encoding on the output. However, apparently, pax headers insist that they are UTF-8, which means that they constrain the output of Bazel  to UTF-8 also. 

Now I see where you are coming from: commons-compress doesn't tell whether there are pax headers and thus it doesn't tell you whether UTF-8 decoding was required. So what you could do is to keep track of `String`s that were emitted by the `Charset`, thus you know which `String`s went through the UTF-8 decoding and thus undo it. Is this correct?

I don't claim any knowledge of either commons-compress or the intricacies of tar files, but `TarArchiveEntry` seems to have some methods that have `Pax` in their name. Maybe those could be used to disambiguate instead of the hack you proposed?

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java:61]: As far as I can tell you can get your hands on *extra* pax headers, but not on the standard ones including the file name. It's just silently updated to the pax header value when encountered.

Comment (tetromino) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java:61]: Note that https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html#tag_20_92_13_03 states that pax headers can be UTF-8 or "binary" (OS-specified encoding).

Is the "binary" option never used in practice?

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java:61]: @tetromino Not in commons-compress, as far as I can tell: https://github.com/apache/commons-compress/blob/3605bc0e27c1f1d27868e7ad5665513835ca7bec/src/main/java/org/apache/commons/compress/archivers/tar/TarUtils.java#L799

Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java:61]: @tetromino WDYT about the above hack with a custom `Charset` @fmeum proposed?

I am kinda drawn to it, because it wouldn't affect anything outside the interpretation of that particular `.tar` file and would allow us to work around the above limitation, but it's messy enough that I don't want to bless it without a second opinion.

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java:61]: @tetromino Friendly ping, let me know whether you want me to implement the hack.

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java:61]: @tetromino Friendly ping

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java]: @lberki Since I wanted to know whether the hack would work, I went ahead and implemented it. It does seem to work and I didn't even have to use a weak cache.

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java]: @lberki @tetromino Friendly ping, I would really like to get this into Bazel 7. We currently have to maintain a number of hacks in rules_go just to be able to extract a Go SDK.

**Comment by lberki** [comment]
@tetromino is not available these days. I don't have any context on this pull request so I wouldn't be able to give it the attention it deserves, but let me try to find someone who has all the bits swapped into their brain.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/DecompressorValue.java
Comment (brandjon) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/DecompressorValue.java:69]: > where strings always use UTF-8

How so? Can't a string literal embed bytes from an arbitrary encoding, and get stored directly in the low bytes of a Java String?

I don't know the significance of this code though. I'm just responding to the claim in the comment.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java
Comment (brandjon) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java]: My gut reaction to the encoding hack was pretty negative. Then I reread the code and description a couple times, and decided that the hack is only ugly because the apache api is insufficiently flexible for bad clients like us that want to do weird things with encodings. I'm ok with the hack provided it is sufficiently explained in comments, and doesn't leak outside this file.

I would like us to be pretty explicit in comments about what the format of the input headers and returned string are. I.e., Javadoc for the class `CompressedTarFunction` or its `decompress()` method, saying that USTAR headers are interpreted as latin-1, PAX headers are interpetted as UTF-8, and both types are turned into Java Strings holding the UTF-8 code units that represent the string content, in order to match our string behavior elsewhere.

One more question. We eventually want to fix Bazel's decoding of Starlark source files to be UTF-8 instead of latin-1. Does anything in this PR make that harder? It seems to me this PR is just preserving an existing behavior (USTAR) while fixing an erroneous one (PAX).

File: src/test/shell/bazel/bazel_workspaces_test.sh
Comment (fmeum) [reply src/test/shell/bazel/bazel_workspaces_test.sh:457]: I added comments to all the tar tests to explain the differences between them.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/DecompressorValue.java
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/DecompressorValue.java]: You are right, I adjusted the comment.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java]: I added more extensive comments, including Javadocs for the class.

> One more question. We eventually want to fix Bazel's decoding of Starlark source files to be UTF-8 instead of latin-1. Does anything in this PR make that harder? It seems to me this PR is just preserving an existing behavior (USTAR) while fixing an erroneous one (PAX).

I don't think that it would make this harder. It does introduce one or two new places in which `ISO_8859_1` would need to be replaced with `UTF-8`, but those are only related to prefix stripping, not the actual entry decoding logic that interfaces with commons-compress and relies on the custom decoder hack.

**Review by brandjon — APPROVED** [review]

**Comment by brandjon** [comment]
Thanks for the additional javadoc/comments.

**Comment by fmeum** [comment]
@bazel-io flag

**Comment by fmeum** [comment]
@brandjon Is this still on track for getting merged?

**Review by lberki — APPROVED** [review]

**Comment by meteorcloudy** [comment]
@brandjon is still fixing some internal breakages.

**Comment by fmeum** [comment]
> @brandjon is still fixing some internal breakages.

Thanks for the update. I didn't know this code is used at Google - isn't it only relevant for external repos?

**Comment by meteorcloudy** [comment]
Might be just some irrelevant breakages, Jon is OOO until next Tuesday, I can help merge this.

**Comment by brandjon** [comment]
Sorry, internal presubmits failed and then I went OOO. Thanks Yun for completing the merge!

**Pull request closed without merging**
